### PR TITLE
Make the WCA SCA ready

### PR DESCRIPTION
**Pull request by viroulep**
Starting in September (on the 14th, 3 days ago actually :p) European payments may require [Strong Customer Authentification](https://stripe.com/en-fr/payments/strong-customer-authentication), and we were not ready for that at all...

We used to use direct charge from the server, which is not possible when a 3D secure challenge is required for the user.
A direct update to the new "checkout" flow would involve asynchronously charging the user, and fulfilling the order through a webhook setup in our Stripe account dashboard.
This would have involved several structural changes and would expose us to losing charge information if our server goes down somewhere in the middle of the process (obviously the charge would still appear on the connected Stripe account, but we may not be able to mark a registration as paid on the website).

Hopefully there is another way to keep a synchronous flow to charge for registration through a combination of Stripe.js, Elements, and the PaymentIntent API (the overall workflow is described [here](https://stripe.com/docs/payments/payment-intents/web-manual)).
It also has the benefit to be very well integrated in the form (I actually added a disclaimer that the card information never reach our server, because it may be confusing).

### New payment flow
TL;DR; deployed to staging, [this](https://staging.worldcubeassociation.org/competitions/ChampagneOpen2019) competition has a linked Stripe account and registration closing in november, so fill free to log in as one of the competitors and play around the payment form!

Basically we include some Stripe javascript in our registration form, which creates an iframe with the inputs for the card information.
Upon clicking the payment button, Stripe generates a PaymentMethod with these information, and POST its id to one of our endpoints (`process_payment_intent`) (we don't leave the page, it's an AJAX query).
Server-side, we can check if any further action is required (eg: 3D secure challenge); if yes, we answer that action is required, if not we place and record the charge (and of course if the payment is refused we record the failure).
Back to client side, if further action is required we tell Stripe to take care of the 3D secure challenge, and Stripe will POST again to the same endpoint the PaymentIntent identifier and the challenge id.
Server-side we attempt to confirm the PaymentIntent with the challenge, if we succeed we record the charge, if not we decline it.
Back to the client, we redirect with a flash message upon success, or we display an error message upon failure.

### Changes in tests

I moved the "controller" specs to "requests" spec, and added some additional test for some specific card failure cases.
By only relying on requests I could only test the 3D secure challenge up to receiving the answer that "action is required".
Therefore I created two extra feature tests to test a successful and failed challenge (they have to actually do some ajax queries, so I extended the timeout and may take like 6-7 seconds each).

### Changes in journaled Stripe charges

I kept the "unknown" status which is the default status before we create any PaymentIntent, and we have a new "payment_intent_registered" status.
I also had to change the type of the `stripe_charge_id`, because it's always a string and not an integer (which mean all our `stripe_charge_id` are either `0` or `nil` currently!!!).

TODO after deploy:
  - [ ] update the default Stripe API version in the dashboard.

### CODE REVIEW

File: WcaOnRails/spec/features/sca_payments_spec.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  within_frame(page.first("#card-element iframe")) do
+    fill_in with: card[:number], name: "cardnumber"
+    fill_in with: "#{card[:exp_month]}#{card[:exp_year]}", name: "exp-date"
+    fill_in with: card[:cvc], name: "cvc"
+    fill_in with: "12345", name: "postal"
+  end
+  click_on "Pay your fees with card"
+  Capybara.using_wait_time(15) do
+    # The 3D secure challenge is in a modal in an iframe in an iframe
+    within_frame(page.find(:xpath, "/html/body/div[1]/iframe")) do
```
Comment (jonatanklosko) [inline WcaOnRails/spec/features/sca_payments_spec.rb]: This is probably as good as `page.find("iframe")`, it's a pity it doesn't have any id : (

File: WcaOnRails/spec/requests/registrations_spec.rb [hunk 2]
```diff
@@ -430,6 +430,212 @@
...
   end
+
+  describe "POST #process_payment_intent" do
+    context "when not signed in" do
+      let(:competition) { FactoryBot.create(:competition, :stripe_connected, :visible, :registration_open, events: Event.where(id: %w(222 333))) }
+      sign_out
+
+      it "redirects to the sign in page" do
+        # 1234 is just some random registration id
+        post registration_payment_intent_path(1234)
```
Comment (jonatanklosko) [inline WcaOnRails/spec/requests/registrations_spec.rb]: What about creating a registration and using its id instead? ^^

File: WcaOnRails/spec/requests/registrations_spec.rb [hunk 3]
```diff
@@ -430,6 +430,212 @@
...
+        it "asks for further action before recording payment" do
+          original_outstanding_fees = registration.outstanding_entry_fees
+          expect(original_outstanding_fees).to be > 0
+          post registration_payment_intent_path(registration.id), params: {
+            payment_method_id: pm.id,
+            amount: original_outstanding_fees.cents,
+          }
+          # The #process_payment_intent endpoint doesn't redirect, it's
+          # the 'register' page which does.
+          expect(registration.reload.outstanding_entry_fees).to eq original_outstanding_fees
```
Comment (jonatanklosko) [inline WcaOnRails/spec/requests/registrations_spec.rb]: Nit: a fancier version to assert that would be:
```ruby
expect {
  post registration_payment_intent_path(registration.id), params: {
    payment_method_id: pm.id,
    amount: original_outstanding_fees.cents,
  }
}.to_not change { registration.reload.outstanding_entry_fees }
```

File: WcaOnRails/spec/requests/registrations_spec.rb [hunk 4]
```diff
@@ -430,6 +430,212 @@
...
+          expect(metadata["metadata"]["wca_id"]).to eq registration.user.wca_id
+        end
+      end
+
+      # The tests below are to test that our endpoint correctly forwards errors,
+      # not to actually test Stripe's correctness...
+      context "rejected credit cards" do
+        it "rejects payment with declined credit card" do
+          original_outstanding_fees = registration.outstanding_entry_fees
+          expect(original_outstanding_fees).to be > 0
```
Comment (jonatanklosko) [inline WcaOnRails/spec/requests/registrations_spec.rb]: Hmm, this is actually pretty repetitive and doesn't really have to do with the tested functionality itself. I'd suggest either making it clear somehow that that's the case (i.e. during creation) or running that in a before block.

File: WcaOnRails/spec/rails_helper.rb [hunk 5]
```diff
@@ -34,7 +34,12 @@
 # To debug feature specs using phantomjs, set `Capybara.javascript_driver = :poltergeist_debug`
 # and then call `page.driver.debug` in your feature spec.
 Capybara.register_driver :poltergeist_debug do |app|
-  Capybara::Poltergeist::Driver.new(app, inspector: true, phantomjs: Phantomjs.path, debug: true)
+  Capybara::Poltergeist::Driver.new(app, inspector: true, phantomjs: Phantomjs.path, debug: true, phantomjs_options: ['--ignore-ssl-errors=yes', '--ssl-protocol=any'])
```
Comment (jonatanklosko) [inline WcaOnRails/spec/rails_helper.rb]: Would be great to have a comment for why we need each of these options =)

File: WcaOnRails/app/controllers/registrations_controller.rb [hunk 6]
```diff
@@ -377,83 +377,121 @@ def register
...
+      wca_id: user.wca_id,
+      email: user.email,
+      competition: competition.name,
+    }
+    begin
+      if params[:payment_method_id]
+        if amount < registration.outstanding_entry_fees.cents
+          render json: { error: { message: t("registrations.payment_form.alerts.amount_too_low") } }
+          return
+        end
```
[the remark this answers is not in the document]
Comment (viroulep) [reply WcaOnRails/app/controllers/registrations_controller.rb]: Unfortunately this won't work, because if there is a `params[:payment_intent_id]` then the amount is not set, yet it's a valid request.

File: WcaOnRails/app/controllers/registrations_controller.rb [hunk 7]
```diff
@@ -377,83 +377,121 @@ def register
...
-    begin
-      charge = Stripe::Charge.create(*stripe_charge_create_args)
-      stripe_charge.update!(
-        status: "success",
-        stripe_charge_id: charge.id,
-      )
-    rescue Stripe::CardError, Stripe::InvalidRequestError => e
+  private def generate_payment_response(registration, intent, stripe_charge)
+    if intent&.status == "requires_action" &&
+       intent.next_action.type == "use_stripe_sdk"
```
[the remark this answers is not in the document]
Comment (viroulep) [reply WcaOnRails/app/controllers/registrations_controller.rb]: I went for `intent &&`, because using `&.` on the second part is useless (because as you noticed a `nil` intent will never pass the left hand side of the `&&`).

File: WcaOnRails/app/controllers/registrations_controller.rb [hunk 8]
```diff
@@ -377,83 +377,121 @@ def register
...
       stripe_charge.update!(
-        status: "failure",
-        error: error_to_s(e),
+        status: "payment_intent_registered",
+        stripe_charge_id: intent.id,
```
[the remark this answers is not in the document]
Comment (viroulep) [reply WcaOnRails/app/controllers/registrations_controller.rb]: Yes and no: at this point the paymentintent has no charge linked, but we should still be able to link this unconfirmed attempt to a stripe action.
Therefore storing the paymentintent id seems like a better solution than storing `nil`.
I've added a comment :)

File: WcaOnRails/app/views/registrations/_payment_form.html.erb [hunk 9]
```diff
@@ -1,52 +1,133 @@
...
       });
-    })();
+    }
+
+    $cardButton.on('click', function(ev) {
+      ev.preventDefault();
+      setSaving();
+      var registration_fees_to_pay = <%= @registration.outstanding_entry_fees.cents %>;
+      var donation = parseFloat($('#payment_donation').val() || 0);
+      var amount = parseFloat(registration_fees_to_pay+donation);
```
[the remark this answers is not in the document]
Comment (viroulep) [reply WcaOnRails/app/views/registrations/_payment_form.html.erb]: Ah right, I'm not sure what I was thinking initially, since this comes from the previous payment form.
I remember trying various amount-formatting library, maybe that was remaining from that.

File: WcaOnRails/config/routes.rb [hunk 10]
```diff
@@ -19,6 +19,7 @@
              end
   end
   post 'registration/:id/refund/:payment_id' => 'registrations#refund_payment', as: :registration_payment_refund
+  post 'registration/:id/process_payment_intent' => 'registrations#process_payment_intent', as: :registration_payment_intent
```
[the remark this answers is not in the document]
Comment (viroulep) [reply WcaOnRails/config/routes.rb]: Ah it was unclear what our convention is, since they are occurrences of both in our routes :p
I'll update the route name!

File: WcaOnRails/app/views/registrations/_payment_form.html.erb [hunk 11]
```diff
@@ -1,52 +1,133 @@
 <% provide(:include_auto_numeric, true) %>
-<%= horizontal_simple_form_for :payment, url: competition_process_payment_path(@competition), html: { id: :form_payment } do |f| %>
+<%= horizontal_simple_form_for :payment, url: "", html: { id: :form_payment } do |f| %>
   <%= render 'entry_fee', label: "Paid", money_amount: @registration.paid_entry_fees %>
   <%= f.input :stripe_token, as: :hidden, input_html: { id: :stripe_token } %>
```
[the remark this answers is not in the document]
Comment (viroulep) [reply WcaOnRails/app/views/registrations/_payment_form.html.erb]: Right!

File: WcaOnRails/spec/features/sca_payments_spec.rb (hunk 1 above)
Comment (viroulep) [reply WcaOnRails/spec/features/sca_payments_spec.rb]: Unfortunately not: there is already an iframe containing the card information which is not removed when the iframe containing the 3D secure modal is mounted.
Since `find` would succeed for the first element found, and since the ajax query takes a bit of time and needs to happen before the modal's iframe is mounted, then it would most likely match the card information iframe.
Being specific about the path here let capybara wait on its own for the completion of the ajax query (up to the timeout), as the path doesn't exist until the query is completed.

File: WcaOnRails/spec/rails_helper.rb (hunk 5 above)
Comment (viroulep) [reply WcaOnRails/spec/rails_helper.rb]: Ah it's just below on the non-debug poltergeist driver!
But I can repeat it :)

File: WcaOnRails/spec/requests/registrations_spec.rb (hunk 2 above)
Comment (viroulep) [reply WcaOnRails/spec/requests/registrations_spec.rb]: It doesn't matter since it shouldn't even look at the registration, but I can!
(The idea was to check it would do the authentication before anything else, and therefore not fail on invalid params but on an invalid authentication)

File: WcaOnRails/app/controllers/registrations_controller.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (jonatanklosko) [reply WcaOnRails/app/controllers/registrations_controller.rb]: Another way would be storing both `stripe_charge_id` and `stript_charge_intent_id`, but either way is fine to me =)

File: WcaOnRails/spec/features/sca_payments_spec.rb (hunk 1 above)
Comment (jonatanklosko) [reply WcaOnRails/spec/features/sca_payments_spec.rb]: Ah gotcha, thanks for explaining that!

File: WcaOnRails/spec/rails_helper.rb (hunk 5 above)
Comment (jonatanklosko) [reply WcaOnRails/spec/rails_helper.rb]: Yeah, I was thinking of a more precise note how each of the phantomjs options helps us ^^

File: WcaOnRails/spec/requests/registrations_spec.rb (hunk 4 above)
Comment (viroulep) [reply WcaOnRails/spec/requests/registrations_spec.rb]: Sure, I've specified an explicit `base_entry_fee_lowest_denomination` when creating the competition.

File: WcaOnRails/app/views/registrations/_payment_form.html.erb [hunk 12]
```diff
@@ -1,52 +1,133 @@
...
+          stripe.handleCardAction(
+            response_json.payment_intent_client_secret
+          ).then(function(result) {
+            if (result.error) {
+              $("#card-error").text(result.error.message);
+              unsetSaving();
+            } else {
+              // The card action has been handled
+              // The PaymentIntent can be confirmed again on the server
+              fetch('<%= registration_payment_intent_path(@registration) %>', {
```
[the remark this answers is not in the document]
Comment (viroulep) [reply WcaOnRails/app/views/registrations/_payment_form.html.erb]: Good point, I've fixed that!

File: WcaOnRails/spec/rails_helper.rb (hunk 5 above)
Comment (viroulep) [reply WcaOnRails/spec/rails_helper.rb]: I've updated the comment!

**Comment by viroulep** [comment]
Will prepare a mail for WCA Staff as this is a major change, then I'll merge and deploy this :)

**Comment by viroulep** [comment]
Oh and here is a screenshot of how the form looks like now in case someone doesn't want to go to staging or an existing unpaid registration:
sca-ui

**Comment by viroulep** [comment]
Looks alive and well :)
